util/logtools/make_timeline.ts

```diff
@@ -0,0 +1,512 @@
+    if (lastEntry.time === entry.time) {
+      if (entry.abilityId && lastEntry.abilityId && entry.abilityId === lastEntry.abilityId)
+        continue;
+    }
+
+    // Ignore targetable lines if not specified
+    if (entry.lineType === 'targetable' && !args['include-targetable'])
+      continue;
+
+    // Find out how long it's been since the last ability.
+    const lineTime = Date.parse(entry.time);
+    const lastTimeDiff = lineTime - lastAbilityTime;
+    const timeInfo = findTimeDifferences(lastTimeDiff);
+
+    // If the user entered phase information,
+    // process it and store it off.
+    const phases: { [name: string]: number } = {};
+    if (args.phase) {
+      for (const phase of args.phase) {
+        const ability = phase.split(':')[0];
+        const time = phase.split(':')[1];
+        if (ability && time)
+          phases[ability] = parseInt(time);
+      }
+    }
+
+    // Set the time, adjusting to phases if necessary.
+    const abilityName = entry.abilityName ?? 'Unknown';
+    const phaseTime = phases[abilityName] ?? timelinePosition;
+    if (
+      !(entry.lineType === 'ability' && Object.keys(phases).includes(abilityName))
+    ) {
+      timelinePosition += timeInfo.diffSeconds;
+    } else if (abilityName && Object.keys(phases).includes(abilityName)) {
+      timelinePosition = phaseTime;
+      delete phases[abilityName];
+    }
+
+    // We're done manipulating time, so save where we are for the next loop.
+    lastAbilityTime = lineTime;
+
+    if (entry.lineType !== 'nameToggle') {
+      const ability = entry.abilityName ?? 'Unknown';
+      const abilityId = entry.abilityId ?? 'Unknown';
+      const combatant = entry.combatant ?? 'Unknown';
+      const newEntry = `${
+        timelinePosition.toFixed(1)
+      } "${ability}" sync / 1[56]:[^:]*:${combatant}:${abilityId}:/`;
+      assembled.push(newEntry);
+    } else {
+      const targetable = entry.targetable ? '--targetable--' : '--untargetable--';
+      const newEntry = `${timelinePosition.toFixed(1)} "${targetable}"`;
+      assembled.push(newEntry);
+    }
+    lastEntry = entry;
+  }
+  return assembled;
+};
+
+const printTimelineToConsole = (entryList: string[]): void => {
+  if (entryList.length > 0)
+    console.log(entryList.join('\n'));
+};
+
+const writeTimelineToFile = (entryList: string[], fileName: string, force: boolean): void => {
+  const flags = force ? 'wx' : 'w';
+  const writer = fs.createWriteStream(fileName, { flags: flags });
+  writer.on('error', (err) => {
+    console.error(err);
+    process.exit(-1);
+  });
+  if (entryList.length > 0) {
+    for (const entry of entryList) {
+      writer.write(entry);
+      writer.write('\n');
+    }
+    writer.close();
+  }
+};
+
+const makeTimeline = async () => {
+  if (args.file) {
+    const collector = await makeCollectorFromPrepass(args.file);
+    if (args['search_fights'] === -1) {
+      TLFuncs.printCollectedFights(collector);
+      process.exit(0);
+    }
```

Comment:
It occurs to me that if we drop support for manual start/end timestamps (or even if we don't) we could eliminate the need to extract lines a second time. I quickly hacked something together on a new branch to test this, and it appears to work okay. We could do something awful like:

```typescript
  if (args.file) {
    const store = (args.search_fights && (args.search_fights > 0)) as boolean;
    const collector = await makeCollectorFromPrepass(args.file, store);
    if (args['search_fights'] === -1) {
      TLFuncs.printCollectedFights(collector);
      process.exit(0);
    }
// etc
```

And have this store an array of lines to each fight in the `EncounterCollector`. It might end up being slightly impractical if someone's logs go over 2 GB, but that's a them problem at that point.

I don't have things set up to benchmark it, but it has to be faster overall since we're completely removing a file access and a loop based on it. The number of lines to loop over in `extractTLEntries` should still end up being the same.